bentoml/_internal/utils/circus.py

```diff
@@ -61,94 +75,160 @@ def create_standalone_arbiter(watchers: list[Watcher], **kwargs: t.Any) -> Arbit
-        pass
-
-
-class PyFileChangeWatcher:
-    def __init__(
-        self,
-        watch_dirs: list[pathlib.Path] | list[str] | None = None,
-    ) -> None:
-        self.mtimes: dict[pathlib.Path, float] = {}
-        if not watch_dirs:
-            watch_dirs = [pathlib.Path.cwd()]
-        self.watch_dirs = [
-            pathlib.Path(d) if isinstance(d, str) else d for d in watch_dirs
-        ]
-        logger.info(f"Watching directories: {', '.join(map(str, self.watch_dirs))}")
-
-    def is_file_changed(self) -> bool:
-        for file in self.iter_files():
-            try:
-                mtime = file.stat().st_mtime
-            except OSError:  # pragma: nocover
-                continue
-
-            old_time = self.mtimes.get(file)
-            if old_time is None:
-                self.mtimes[file] = mtime
-                continue
-            elif mtime > old_time:
-                display_path = str(file)
-                try:
-                    display_path = str(file.relative_to(pathlib.Path.cwd()))
-                except ValueError:
-                    pass
-                message = "Detected file change in '%s'"
-                logger.warning(message, display_path)
-                return True
-        return False
+        self.exit_event.set()
+        self.thread.join()
+
+    def _display_path(
+        self, change: t.Set[FileChange]
+    ) -> tuple[str, str]:  # pragma: no cover
+        type_change, path = change.pop()
+        try:
+            return (
+                type_change.raw_str(),
+                f"'{Path(path).relative_to(self._cwd)}'",
+            )
+        except ValueError:
+            return type_change.raw_str(), f"'{path}'"
+
 
-    def reset(self) -> None:
-        self.mtimes = {}
+class BentoFilter:
+    """
+    A Python filter that respect bentofile.yaml and .bentoignore
+    Currently ignoring `bentofile.exclude` fields.
+    """
 
-    def iter_files(self) -> t.Iterator[pathlib.Path]:
-        for reload_dir in self.watch_dirs:
-            for path in list(reload_dir.rglob("*.py")):
-                yield path.resolve()
+    def __init__(self, path: PathType):
+        if isinstance(path, Path):
+            path = path.__fspath__()
+
+        self._fs = fs.open_fs(str(path))
+
+        if self._fs.exists("bentofile.yaml"):
+            with self._fs.open("bentofile.yaml", "r", encoding="utf-8") as f:
+                build_config = BentoBuildConfig.from_yaml(f.read()).with_defaults()
+        else:
+            # We just need include and exclude fields to create spec.
+            # Hence, service can be an empty string.
+            build_config = BentoBuildConfig(service="").with_defaults()
+
+        include = build_config.include
+        exclude = build_config.exclude
+
+        # bentoml/_internal/bento/bento.py#L191
+        self.include_spec = PathSpec.from_lines("gitwildmatch", include)
+        self.exclude_spec = PathSpec.from_lines("gitwildmatch", exclude)
```

Comment:
Hm, should we factor this out into its own method? Or maybe attach it to the `BentoBuildConfig`?

Reply:
Something like `BentoBuildConfig.includes_file(filename)` maybe would be nice.

Reply:
sounds good